src/lib.rs

```diff
@@ -29,3 +35,125 @@ mod violation;
+/// let owned_element: OwnedElement = 4.into();
+/// let ion_schema_element: IonSchemaElement = (&owned_element).into();
+///
+/// // create an IonSchemaElement for document type based on vector of owned elements
+/// let document: IonSchemaElement = IonSchemaElement::Document(vec![owned_element]);
+/// ```
+#[derive(Debug, Clone)]
+pub enum IonSchemaElement {
+    SingleElement(OwnedElement),
+    Document(Vec<OwnedElement>),
+}
+
+impl IonSchemaElement {
+    pub fn as_element(&self) -> Option<&OwnedElement> {
+        match self {
+            IonSchemaElement::SingleElement(element) => Some(element),
+            IonSchemaElement::Document(_) => None,
+        }
+    }
+
+    pub fn as_document(&self) -> Option<&Vec<OwnedElement>> {
+        match self {
+            IonSchemaElement::SingleElement(_) => None,
+            IonSchemaElement::Document(document) => Some(document),
+        }
+    }
+
+    fn expect_element_of_type(
+        &self,
+        constraint_name: &str,
+        types: &[IonType],
+    ) -> Result<&OwnedElement, Violation> {
+        match self {
+            IonSchemaElement::SingleElement(element) => {
+                if !types.contains(&element.ion_type()) || element.is_null() {
+                    // If it's an OwnedElement but the type isn't one of `types`,
+                    // return a Violation with the constraint name.
+                    return Err(Violation::new(
+                        constraint_name,
+                        ViolationCode::TypeMismatched,
+                        &format!("expected {:?} but found {}", types, element.ion_type()),
+                    ));
+                }
+                // If it's an OwnedElement of an expected type, return a ref to it.
+                Ok(element)
+            }
+            IonSchemaElement::Document(_) => {
+                // If it's a Document, return a Violation with the constraint name
+                Err(Violation::new(
+                    constraint_name,
+                    ViolationCode::TypeMismatched,
+                    &format!("expected {:?} but found document", types),
+                ))
+            }
+        }
+    }
+}
+
+impl Display for IonSchemaElement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match self {
+            IonSchemaElement::SingleElement(element) => {
+                write!(f, "{}", element)
+            }
+            IonSchemaElement::Document(document) => {
+                write!(f, "/* Ion document */ ")?;
+                let mut peekable_itr = document.iter().peekable();
+                while peekable_itr.peek() != None {
+                    let value = peekable_itr.next().unwrap();
+                    write!(f, "{}", value)?;
+                    if peekable_itr.peek() != None {
+                        write!(f, " ")?;
+                    }
+                }
+                write!(f, " /* end */")
```

Comment:
minor suggestion—you can eliminate the need for peeking by dropping the space before the "/* end */" comment, which will allow you to simplify the code to something like this:
```rust
write!(f, "/* Ion document */ ")?;
for value in document {
    write!(f, "{} ", value)?;
}
write!(f, "/* end */")
```
(You should double check if you choose to do this—I'm not sure off the top of my head whether it should be `document` or `&document`.)